File: LeetCode/233-Number of Digit One/solution.py
#!/usr/local/bin/python3
# coding: UTF-8
# Author: David
# Email: youchen.du@gmail.com
# Created: 2017-03-01 22:03
# Last modified: 2017-03-01 22:03
# Filename: solution.py
# Description:


class Solution(object):
    def countDigitOne(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n <= 0:
            return 0
        elif n < 10:
            return 1
        from math import log10
        s = str(n)
        k = int(s[0])
        r = int(s[1:])
        m = int(log10(n))
        ones = 10 ** m
        cnt = k * self.countDigitOne(ones - 1) + self.countDigitOne(r)
        if n >= 2 * ones:
            cnt += ones
        else:
            cnt += n - ones + 1
        return cnt

File: LeetCode/233-Number of Digit One/solution.cpp
// Author: David
// Email: youchen.du@gmail.com
// Created: 2017-03-02 09:33
// Last modified: 2017-03-02 09:33
// Filename: solution.cpp
// Description:
class Solution {
public:
    int countDigitOne(int n) {
        int l, r = n, x = 1;
        int ans = 0;
        while(r > 0)
        {
            l = r % 10;
            r = r / 10;
            ans += r * x;
            if(l == 1)
                ans += 1 + n % x;
            else if(l > 1)
                ans += x;
            x = x * 10;
        }
        return ans;
    }
};
